File: c++/0347-top-k-frequent-elements/solution.cpp
#include <queue>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map;
        for (int n : nums) {
            ++map[n];
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> queue;
        for (auto p : map) {
            queue.push({p.second, p.first});

            if (queue.size() > k) {
                queue.pop();
            }
        }

        vector<int> result;
        while (!queue.empty()) {
            result.push_back(queue.top().second);
            queue.pop();
        }

        return result;
    }
};

File: java/0347-top-k-frequent-elements/Solution.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int n : nums) {
            map.put(n, map.getOrDefault(n, 0) + 1);
        }

        PriorityQueue<Map.Entry<Integer, Integer>> priorityQueue = new PriorityQueue<>(
                (a, b) -> Integer.compare(a.getValue(), b.getValue())
        );
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            priorityQueue.add(entry);

            if (priorityQueue.size() > k) {
                priorityQueue.poll();
            }
        }

        int[] result = new int[k];
        for (int i = k - 1; !priorityQueue.isEmpty(); --i) {
            result[i] = priorityQueue.poll().getKey();
        }

        return result;
    }
}
